src/python/pants/engine/legacy/structs.py

```diff
@@ -29,6 +29,20 @@ class TargetAdaptor(StructWithDeps):
   Extends StructWithDeps to add a `dependencies` field marked Addressable.
   """
 
+  def _coerce_key_values(self, key, value):
+    # TODO: This method is overridden from `StructWithDeps` to remove elements which cause
+    # `stable_json_sha1()` to fail with a cycle detection. Since some python targets are only mapped
+    # to `TargetAdaptor` (and not `PythonTargetAdaptor`), we check every single target for a
+    # `requirements` kwarg, which is fine for now.
+    key, value = super()._coerce_key_values(key, value)
```

Comment:
I'm not sure from reading - does this cover all actual key-values of the `Target`, or just the ones explicitly listed in BUILD files?

In particular, if we change a default value in pants, or set a default with a flag or something, and the target doesn't set it, will the fingerprint change? Presumably it _should_, right?

Reply:
This covers all the key-values that are provided as `field_adaptors`, I believe, essentially because we're getting this info from a `TargetAdaptor`, which will explicitly only use `HydrateableField` `@union` members as the `_kwargs` provided by `Struct` -- see `hydrate_targets`: https://github.com/pantsbuild/pants/blob/1988587507cb2cb21efffac01ac9a8c3ebf291a5/src/python/pants/engine/legacy/graph.py#L497-L509

> In particular, if we change a default value in pants, or set a default with a flag or something, and the target doesn't set it, will the fingerprint change? Presumably it *should*, right?

Short answer: no, this fingerprint will not necessarily change, and yes, I absolutely think it should before we merge this.

Longer answer: Not all the things that contribute to `Target#fingerprint` do _not_ transfer to a `TargetAdaptor` (which subclasses `StructWithDeps`, which subclasses `Struct`), *only* the things which are marked as hydrateable fields. As you imply, this is possibly not what we want at all for this purpose, and my use of `TargetAdaptor` here might not be correct.

@stuhood do you have any insight on how to bridge this? Does it break the v2 build graph traversal model if we're able to get an instance of a real `Target` in order to get a more normal fingerprint (i.e. a fingerprint containing everything the target payload does)? Are the `Payload`/`Target` concepts necessarily v1-only/requiring a v1 build graph, or is it possible to avoid having to reimplement payloads for all targets? I would love to pair on this or make an issue as necessary.

Reply:
Can we remove the fingerprints from the output before merging this?
It feels like target_type is an obviously useful thing to include; it's less obvious that these fingerprints are things we should be exposing as-is, but we can always add them in the future if we need to / firm them up :)